src/buckets.spec.ts

```diff
@@ -0,0 +1,200 @@
+    const id = ThreadID.fromRandom()
+    const db = new Client(ctx)
+    // @todo: Warning, this is a hack!
+    ctx.host = client.serviceHost
+    ;(db as any).config = ctx
+    await db.newDB(id.toBytes())
+    ctx = ctx.withThread(id)
+  })
+
+  it('should init a new bucket', async () => {
+    // Check that we're empty
+    const list = await client.list(ctx)
+    expect(list).to.have.length(0)
+    // Now initialize a bucket
+    buck = await client.init('mybuck', ctx)
+    expect(buck).to.have.ownProperty('root')
+    expect(buck.root).to.have.ownProperty('key')
+    expect(buck.root).to.have.ownProperty('path')
+    expect(buck.root).to.have.ownProperty('createdat')
+    expect(buck.root).to.have.ownProperty('updatedat')
+  })
+
+  it('should list buckets', async () => {
+    const roots = await client.list(ctx)
+    expect(roots).to.have.length(1)
+    const root = roots[0]
+    expect(root).to.have.ownProperty('key', buck.root?.key)
+    expect(root).to.have.ownProperty('path', buck.root?.path)
+    expect(root).to.have.ownProperty('createdat', buck.root?.createdat)
+    expect(root).to.have.ownProperty('updatedat', buck.root?.updatedat)
+  })
+
+  it('should list empty bucket content at path', async () => {
+    // Mostly empty
+    const res = await client.listPath(buck.root?.key || '', '', ctx)
+    expect(res).to.have.ownProperty('root')
+    expect(res.root).to.not.be.undefined
+    expect(res.item?.isdir).to.be.true
+    // @todo: Should we rename itemsList to just be items?
+    expect(res.item?.itemsList).to.have.length(0)
+  })
+
+  it('should push data from filesystem on node', async function () {
+    if (isBrowser) return this.skip()
+    const pth = path.join(__dirname, '..', 'testdata')
+    fileSize = fs.statSync(path.join(pth, 'file1.jpg')).size
+    let stream = fs.createReadStream(path.join(pth, 'file1.jpg'))
+    const rootKey = buck.root?.key || ''
+    let length = 0
+
+    // Bucket path
+    const res = await client.pushPath(rootKey, 'dir1/file1.jpg', stream, ctx, {
+      progress: (num) => (length = num || 0),
+    })
+    expect(length).to.equal(fileSize)
+    expect(res.path).to.not.be.undefined
+    expect(res.root).to.not.be.undefined
+
+    // Nested bucket path
+    stream = fs.createReadStream(path.join(pth, 'file2.jpg'))
+    const { root } = await client.pushPath(rootKey, 'path/to/file2.jpg', stream, ctx)
+    expect(root).to.not.be.undefined
+
+    // Root dir
+    const rep = await client.listPath(rootKey, '', ctx)
+    expect(rep.item?.isdir).to.be.true
+    expect(rep.item?.itemsList).to.have.length(2)
+  })
+
+  it('should push data from file API in browser', async function () {
+    if (isNode) return this.skip()
+    const parts = [
+      new Blob(['you construct a file...'], { type: 'text/plain' }),
+      ' Same way as you do with blob',
+      new Uint16Array([33]),
+    ]
+    // Construct a file
+    const file = new File(parts, 'file1.txt')
```

Comment:
This is the equivalent of getting a file object from a drag and drop event or a file picking in the browser, which is likely a pretty standard use case. The nice thing here is we can essentially stream the file to our grpc method, but it feels like a standard promise-based rest call to the developer.